### ISSUE TITLE
Swift - parser rules with Optional parameters

### ISSUE THREAD

**Issue by MutatingFunk** [antlr/antlr4#2148]
Example code:
```
rule2: ruleName[nil]
rule1[param: Type?]: rule0[param]
rule0[param: Type?]: Rule Syntax
```

Expected behaviour:
`nil` can be safely passed to a rule with an optional argument.

_(incidentally, I'd also expect the rule to be called as `ruleName[param: nil]`, allowing declaration as `ruleName[_ param: Type?]` to remove the argument label, but I can live without if there are technical/consistency reasons for this)_

Actual behaviour:
Passing `nil` will cause a runtime crash. The generated parameter is actually `Type?!`, an Optional wrapped in an ImplicitlyUnwrappedOptional (IUO), and the nil is inferred to be `IUO<Optional<Type>>.none`, instead of the expected `IUO(Optional<Type>.none)`.

Attempted workaround:
`rule2: rule1[nil]` → `rule2: rule1[.some(nil)]`
This still fails, but later. In the generated code, it seems the IUO.some(nil) is somehow transformed to IUO.none when passed to next rule.

Failing this, it seems there is no workaround to use optionals as parameters, without the fix of Antlr being modified to generate non-IUO parameters. Preferably, it would generate raw `Type` parameters, but if consistency is needed, perhaps `Type?` would be workable?

**Comment by ewanmellor** [comment]
@MutatingFunk You'll need to leave the optionals out.  Rule parameters are going to be generated as `Type!`, as you've already figured out.

I looked at fixing this recently, but since it's a concept that we don't have in Antlr's Java-based roots, it was a really painful and invasive change, so I gave up.

Using `Type!` is a compromise, but it's been one that I haven't had too many problems dealing with (for my use cases, obviously).
